diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index d64ed3aa9430..7296367e1ee5 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -8283,6 +8283,46 @@ static void kvm_sched_yield(struct kvm *kvm, unsigned long dest_id)
 		kvm_vcpu_yield_to(target);
 }
 
+static void cat_setup(void* garbage) {
+    //printk(KERN_INFO "kvm: KVM_HC_CAT_SETUP called\n");
+
+    // setup CBM1
+    asm __volatile__(
+		    "movl $0x0c91, %%ecx\n\t"
+		    "movl $0, %%edx\n\t"
+		    "movl $0x0f, %%eax\n\t"
+		    "wrmsr\n\t"
+		    :
+		    :
+		    : "%rcx", "%rdx", "%rax");
+
+    // setup CBM2
+    asm __volatile__(
+		    "movl $0x0c92, %%ecx\n\t"
+		    "movl $0, %%edx\n\t"
+		    "movl $0x0f0, %%eax\n\t"
+		    "wrmsr\n\t"
+		    :
+		    :
+		    : "%rcx", "%rdx", "%rax");
+
+}
+
+static void cat_alloc(void* cos) {
+    unsigned int cos_num = (unsigned int)(unsigned long int)cos;
+    //printk(KERN_INFO "kvm: KVM_HC_CAT_ALLOC called with %ld\n",cos_num);
+
+    // set COS
+    asm __volatile__(
+		    "movl $0x0c8f, %%ecx\n\t"
+		    "rdmsr\n\t"
+		    "movl %0, %%edx\n\t"
+		    "wrmsr\n\t"
+		    :
+		    : "r"(cos_num)
+		    : "%rax", "%rcx", "%rdx");
+}
+
 int kvm_emulate_hypercall(struct kvm_vcpu *vcpu)
 {
 	unsigned long nr, a0, a1, a2, a3, ret;
@@ -8343,6 +8383,18 @@ int kvm_emulate_hypercall(struct kvm_vcpu *vcpu)
 			break;
 
 		kvm_sched_yield(vcpu->kvm, a0);
+		ret = 0;
+		break;
+	case KVM_HC_CAT_SETUP:
+		// set up the CBMs on calling core
+		smp_call_function_single(vcpu->cpu, cat_setup, NULL, 1);
+
+		ret = 0;
+		break;
+	case KVM_HC_CAT_ALLOC:
+		// switch COS on calling core
+		smp_call_function_single(vcpu->cpu, cat_alloc, (void*)a0, 1);
+
 		ret = 0;
 		break;
 	default:
diff --git a/include/uapi/linux/kvm_para.h b/include/uapi/linux/kvm_para.h
index 8b86609849b9..3720efbf716f 100644
--- a/include/uapi/linux/kvm_para.h
+++ b/include/uapi/linux/kvm_para.h
@@ -27,9 +27,13 @@
 #define KVM_HC_MIPS_EXIT_VM		7
 #define KVM_HC_MIPS_CONSOLE_OUTPUT	8
 #define KVM_HC_CLOCK_PAIRING		9
-#define KVM_HC_SEND_IPI		10
+#define KVM_HC_SEND_IPI		        10
 #define KVM_HC_SCHED_YIELD		11
 
+/* Hypercalls for cache partitioning */
+#define KVM_HC_CAT_SETUP                12
+#define KVM_HC_CAT_ALLOC                13
+
 /*
  * hypercalls use architecture specific
  */
